File: files/000-099/077/CH-077.R
library(tidyverse)
library(readxl)

path = "files/CH-77 Character-Based Rhombus.xlsx"
test = read_xlsx(path, range = "C2:Q16", col_names = FALSE) %>%
  map_df(~replace_na(.x, "_")) %>%
  unite("rhombus", everything(), sep = "")

draw_rhombus = function(diag) {
  if (diag %% 2 == 0) {
    stop("diag must be an odd number")
  }
  
  rhombus = matrix(NA, nrow = diag, ncol = diag)
  seq = seq(1, diag, by = 2)
  rev_seq = rev(seq)[-1] 
  seq = c(seq, rev_seq)
  
  for (i in 1:diag) {
    rhombus[i, 1:seq[i]] = "*"
  }
  
  rhombus = as_tibble(rhombus) %>%
    mutate_all(as.character) %>%
    unite("rhombus", everything(), sep = "", na.rm = TRUE) %>%
    mutate(rhombus = str_pad(rhombus, diag, side = "both", pad = "_"))
  
  return(rhombus)
}

result = draw_rhombus(15)

identical(result, test)
#> [1] TRUE

File: files/000-099/077/CH-077.py
import pandas as pd
import numpy as np

path = "CH-77 Character-Based Rhombus.xlsx"
test = pd.read_excel(path, header=None, usecols="C:Q", skiprows=1, nrows=16)
test = test.fillna("_")
test = test.apply(lambda x: "".join(x), axis=1)

def draw_rhombus(diag):
    if diag % 2 == 0:
        raise ValueError("diag must be an odd number")
    
    rhombus = np.full((diag, diag), "_")
    seq = np.arange(1, diag+1, 2)
    rev_seq = np.flip(seq)[1:]
    seq = np.concatenate((seq, rev_seq))
    
    for i in range(diag):
        rhombus[i, diag//2-seq[i]//2:diag//2+seq[i]//2+1] = "*"

    rhombus = pd.DataFrame(rhombus)
    rhombus = rhombus.apply(lambda x: "".join(x), axis=1)

    return rhombus

result = draw_rhombus(15)

print(result.equals(test)) # True